Backpack/src/main/java/net/skyscanner/backpack/contentbubble/BpkContentBubble.kt

```diff
@@ -0,0 +1,224 @@
+import net.skyscanner.backpack.util.use
+
+/**
+ * [BpkContentBubble] is designed to render a single item inside a "bubble".
+ * Think of message apps and how they show content inside text bubbles.
+ *
+ * No background or padding should be set directly to this view, set it directly its
+ * child.
+ */
+open class BpkContentBubble @JvmOverloads constructor(
+  context: Context,
+  attrs: AttributeSet? = null,
+  defStyleAttr: Int = 0,
+  private val pointerDrawable: Drawable = ContextCompat.getDrawable(context, R.drawable.content_bubble_default_pointer)!!,
+  private val cornerRadiusDrawable: Drawable = ContextCompat.getDrawable(context, R.drawable.content_bubble_default_radii)!!
+) : FrameLayout(context, attrs, defStyleAttr) {
+
+  enum class PointerPosition(internal val id: Int, internal val offset: Float) {
+    START(0, 0.25f),
+    MIDDLE(1, 0.5f),
+    END(2, 0.75f)
+  }
+
+  private lateinit var mask: Bitmap
+  private lateinit var maskCanvas: Canvas
+
+  private lateinit var radiusMask: Bitmap
+  private lateinit var radiusMaskCanvas: Canvas
+
+  private val clipRect = RectF()
+  private val porterDuffXfermode = PorterDuffXfermode(PorterDuff.Mode.DST_OUT)
+  private val paint = Paint().apply {
+    xfermode = porterDuffXfermode
+    isAntiAlias = true
+  }
+
+  private var _fitContent = false
+  /**
+   * When fitContent is true the view will ensure extra space is reserved for the bubble pointer to
+   * ensure all content is visible.
+   */
+  var fitContent: Boolean
+    get() = _fitContent
+    set(value) {
+      _fitContent = value
+      requestLayout()
+    }
+
+  private var _round = false
+  /**
+   * Specify if corner radius should be added or not.
+   */
+  var round: Boolean
+    get() = _round
+    set(value) {
+      _round = value
+      requestLayout()
+    }
+
+  private var _pointerPosition = PointerPosition.MIDDLE
+  /**
+   * Specify where the pointer should be rendered.
+   *
+   * @see [PointerPosition]
+   */
+  var pointerPosition: PointerPosition
+    get() = _pointerPosition
+    set(value) {
+      _pointerPosition = value
+      requestLayout()
+    }
+
+  init {
+    this.initialize(attrs, defStyleAttr)
+  }
+
+  private fun initialize(attrs: AttributeSet?, defStyleAttr: Int) {
+    context.theme.obtainStyledAttributes(attrs, R.styleable.BpkContentBubble, defStyleAttr, 0)
+      ?.use {
+        _fitContent = it.getBoolean(R.styleable.BpkContentBubble_contentBubbleFitContent, fitContent)
+        _round = it.getBoolean(R.styleable.BpkContentBubble_contentBubbleRound, round)
+
+        _pointerPosition = it.getInt(R.styleable.BpkContentBubble_contentBubblePointerPosition, pointerPosition.id)
+          .let(::mapXmlToPointerPosition) ?: pointerPosition
+      }
+
+    background = null
+    setWillNotDraw(false)
+    setPadding(0, 0, 0, 0)
+    // PorterDuffXfermode doesn't work in some devices when hardware acceleration is on
+    setLayerType(LAYER_TYPE_SOFTWARE, null)
+
+    pointerDrawable.apply {
```

Comment:
We don't need to store the drawable. Statically-stored bitmap will be sufficient.